src/serializer/ResidualHeapVisitor.js

```diff
@@ -487,30 +495,90 @@ export class ResidualHeapVisitor {
     });
   }
 
-  visitRoots(): void {
-    this.visitGenerator(this.realmGenerator);
+  visitRoots(additionalFunctionValuesAndEffects: Map<FunctionValue, Effects>): void {
+    let generator = this.realm.generator;
+    invariant(generator);
+    this.visitGenerator(generator);
     for (let [, moduleValue] of this.modules.initializedModules) this.visitValue(moduleValue);
-    // Do a fixpoint over all pure generator entries to make sure that we visit
-    // arguments of only BodyEntries that are required by some other residual value
-    let oldDelayedEntries = [];
-    while (oldDelayedEntries.length !== this.delayedVisitGeneratorEntries.length) {
-      oldDelayedEntries = this.delayedVisitGeneratorEntries;
-      this.delayedVisitGeneratorEntries = [];
-      for (let { generator, entry } of oldDelayedEntries)
-        this._withScope(generator, () => {
-          generator.visitEntry(entry, this.createGeneratorVisitCallbacks(generator));
-        });
-    }
+    this.realm.evaluateAndRevertInGlobalEnv(() => {
+      if (additionalFunctionValuesAndEffects.size) {
+        for (let [functionValue, effects] of additionalFunctionValuesAndEffects.entries()) {
+          let [r, g, ob, pb: Map<PropertyBinding, void | Descriptor>, co] = effects;
+          // Need to do this fixup because otherwise we will skip over this function's
+          // generator in the _getTarget scope lookup
+          g.parent = functionValue.parent;
+          functionValue.parent = g;
+          // result -- ignore TODO: return the result from the function somehow
+          // Generator -- visit all entries
+          // Bindings -- only need to serialize bindings if they're captured by some nested function ??
+          //          -- need to apply them and maybe need to revisit functions in ancestors to make sure
+          //          -- we don't overwrite anything they capture
+          //          -- TODO: deal with these properly
+          // PropertyBindings -- visit any property bindings that aren't to createdobjects
+          // CreatedObjects -- should take care of itself
+          this.realm.applyEffects([r, new Generator(this.realm), ob, pb, co]);
+          // Allows us to emit function declarations etc. inside of this additional
+          // function instead of adding them at global scope
+          this.baseContext = functionValue;
+          let visitPropertiesAndBindings = () => {
+            for (let propertyBinding of pb.keys()) {
+              let binding: PropertyBinding = ((propertyBinding: any): PropertyBinding);
+              let object = binding.object;
+              // These should be in the generator so we can skip them
+              if (object.isPartialObject()) continue;
+              invariant(object instanceof ObjectValue);
+              if (co.has(object)) continue; // Created Object's binding
+              if (object.refuseSerialization) continue; // modification to internal state
```

Comment:
So you skip objects marked as `refuseSerialization` as a root. Just to make should, and to indicate intent, also add an invariant to `visitValue` that we'll never visit a value that has `refuseSerialization` set.